### PR TITLE
[21304] Add notification for removed watchers

### PR DESCRIPTION
**Pull request by vlad-tim**
From [Issue 21304](https://community.openproject.com/projects/openproject/work_packages/21304):

> As a OpenProject user
I want OpenProject to automatically send emails to removed watchers
so that I can easily inform these users about new information.

This PR will implement email notifications to removed watchers by analogy with existing emails to added watchers.

### CODE REVIEW

File: lib/services/remove_watcher.rb [hunk 1]
```diff
@@ -36,6 +36,11 @@ def run(success: -> {}, failure: -> {})
     if @work_package.watcher_users.include?(@user)
       @work_package.watcher_users.delete(@user)
```
Comment (ulferts) [inline lib/services/remove_watcher.rb:38]: If the watcher for the user where to be retrieved first, and then destroyed, that destroyed object could then be passed on through the notification without the need to bloat up the interface from three parameters `watcher_toggled`, `watcher`, `watcher_setter` to five. 
The same would then be true for the `WatcherNotificationMailer.handle_watcher_toggle` method.

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -352,6 +345,21 @@ def set_work_package_headers(work_package)
...
+
+      set_work_package_headers(work_package)
+      message_id work_package, user
+      references work_package, user
+
+      with_locale_for(user) do
+        mail to: user.mail, subject: subject_for_work_package(work_package)
+      end
+    end
+  end
```
Comment (ulferts) [inline app/mailers/user_mailer.rb]: 👍 for refactoring this into a separate method

File: app/mailers/user_mailer.rb [hunk 3]
```diff
@@ -75,18 +75,11 @@ def work_package_updated(user, journal, author = User.current)
...
   def work_package_watcher_added(work_package, user, watcher_setter)
-    User.execute_as user do
-      @issue = work_package
-      @watcher_setter = watcher_setter
-
-      set_work_package_headers(work_package)
-      message_id work_package, user
-      references work_package, user
+    work_package_watcher_toggled(work_package, user, watcher_setter)
+  end
```
Comment (ulferts) [inline app/mailers/user_mailer.rb]: You could use `alias` here and for the remove method as well.

File: app/models/watcher_notification_mailer.rb [hunk 4]
```diff
@@ -49,25 +49,25 @@ def other_jobs_queued?(work_package)
                          "%NotificationJob%journal_id: #{work_package.journals.last.id}%").exists?
     end
 
-    def notify_about_watcher_added?(watcher, watcher_setter)
-      return false if notify_about_self_watching?(watcher, watcher_setter)
+    def notify_about_watcher_added?(user, watcher_setter, watchable)
```
Comment (ulferts) [inline app/models/watcher_notification_mailer.rb]: The method could be renamed now( and was misnamed before) as it does not check for whether the user wants to be informed about a watcher being added but, regardless of the action, about events on the watchable.

File: app/views/user_mailer/work_package_watcher_removed.html.erb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+++#%>
+<%= t(:text_work_package_watcher_removed, id: "##{@issue.id}", watcher_setter: @watcher_setter) %>
+<hr />
+<%= render partial: 'issue_details', locals: { issue: @issue } %>
+<p>
+  <%= format_text(t(:text_latest_note, note: last_issue_note(@issue)),
+                  only_path: false,
+                  object: @issue,
+                  project: @issue.project) %>
+</p>
```
Comment (ulferts) [inline app/views/user_mailer/work_package_watcher_removed.html.erb]: The only difference between the 'work_package_watcher_added' view and this one is the translation. Extracting the shared code into a partial with the translation key being a parameter would remove the duplication.

File: app/views/user_mailer/work_package_watcher_removed.text.erb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+See docs/COPYRIGHT.rdoc for more details.
+
+++#%>
+
+<%= t(:text_work_package_watcher_removed, id: "##{@issue.id}", watcher_setter: @watcher_setter) %>
+
+----------------------------------------
+<%= render partial: 'issue_details', locals: { issue: @issue } %>
+<%= t(:text_latest_note, note: last_issue_note(@issue)) %>
```
Comment (ulferts) [inline app/views/user_mailer/work_package_watcher_removed.text.erb]: Same as above.

**Review by ulferts — CHANGES_REQUESTED** [review]
Looks good functionality wise. The notification is sent as desired when the watcher is removed. 

I believe that some parts of the code need improvement before the PR can be merged. Please see my remarks within the code for the most part but I would want to point out two areas explicitly where improvement is needed:
* The added/altered functionality needs to be tested.
* Not relying on the watcher object leads to a couple of interface becoming bloated. The reason for that is to be found in the `RemoveWatcher` job where the watcher is not retrieved which then requires to pass all the parameters that could be passed within the one watcher object.

**Comment by vlad-tim** [comment]
@ulferts thanks for your review, I will let you know when the improvements are made.

**Comment by vlad-tim** [comment]
@ulferts please take another look at my PR. 

Several things I'd like to note explicitly in regards to it:
* The tests have been fixed and extended
* The `watcher` argument is back for `RemoveWatcher` and `CreateWatcher`, 
`WatcherNotificationMailer.handle_watcher_toggle`, and
`Notifications.subscribe('watcher_toggled')`, reducing the number of
arguments in those methods
* In `DeliverWatcherNotificationJob` it still makes sense to replace `watcher` with
  `watchable` because
* 1. `watcher.user.id` is passed as a 2nd argument anyway, so the interface
doesn't become more bloated
* 2. when the job is related to **removing** the watcher, the `watcher` object is
already deleted. One possible solution would be reconstructing it from the attributes just for
the sake of extracting `watchable` and `user` back, but it's arguably not a
perfect approach
* At first I created partials for the mail view, but then realized that the i18n key can be set inside `user_mailer.work_package_watcher_toggled`, removing the need for extra views or particles
* The number of arguments in several methods can be further reduced by eliminating `is_watching` argument, but there is a trade-off - it would lead to necessity to add more classes/files/views with similar looks

In its current form my PR contains a smaller number of changes in production code than the first version to achieve the same goal.

**Comment by lindenthal** [comment]
closed due to inactivity. Please reopen if you want to continue on this PR.

**Pull request closed without merging**